src/backend/utils/misc/queryenvironment.c

```diff
@@ -978,8 +1111,469 @@ void ENRDropEntry(Oid id)
+			continue;
+		}
+
+		/* ROLLBACK block. */
+		in_enr_rollback = true;
+
+		for (int i = 0; i < ENR_CATTUP_END; i++)
+		{
+			List *uncommitted_cattups = enr->md.uncommitted_cattups[i];
+			ListCell *lc2;
+
+			foreach(lc2, uncommitted_cattups)
+			{
+				ENRUncommittedTuple uncommitted_tup = (ENRUncommittedTuple) lfirst(lc2);
+				bool skip_cache_inval = false;
+
+				if (!OidIsValid(uncommitted_tup->catalog_oid) || !HeapTupleIsValid(uncommitted_tup->tup))
+					elog(ERROR, "ROLLBACK failed! Invalid temp table entries were found.");
+
+				if (uncommitted_tup->catalog_oid == IndexRelationId && uncommitted_tup->optype == ENR_OP_ADD)
+				{
+					/*
+					* We skip invalidating IndexRelationId on drop (ie committed table, index created in transaction) 
+					* because it will be taken care of when the index itself is dropped - otherwise we risk
+					* throwing an error because the entry is already wiped away.
+					*/
+					Form_pg_index idx_form = (Form_pg_index) GETSTRUCT(uncommitted_tup->tup);
+					EphemeralNamedRelation tmp_enr = get_ENR_withoid(queryEnv, idx_form->indrelid, ENR_TSQL_TEMP);
+					if (tmp_enr && tmp_enr->md.is_committed)
+						skip_cache_inval = true;
+				}
+
+				rel = relation_open(uncommitted_tup->catalog_oid, AccessExclusiveLock);
+
+				switch(uncommitted_tup->optype)
+				{
+					/* Peform the opposite operation as the saved optype. */
+					case ENR_OP_ADD:
+						_ENR_tuple_operation(rel, uncommitted_tup->tup, ENR_OP_DROP, skip_cache_inval);
+						break;
+					case ENR_OP_UPDATE:
+						ENRupdateTuple(rel, uncommitted_tup->tup);
+						break;
+					case ENR_OP_DROP:
+						_ENR_tuple_operation(rel, uncommitted_tup->tup, ENR_OP_ADD, skip_cache_inval);
+						break;
+				}
+				relation_close(rel, AccessExclusiveLock);
+			}
+		}
+
+		in_enr_rollback = false;
+
+		ENRDeleteUncommittedTupleData(InvalidSubTransactionId, enr);
+	}
+
+	/* Finally, clean up any ENRs that must be deleted. */
+	foreach(lc, enrs_to_drop)
+	{
+		EphemeralNamedRelation enr = (EphemeralNamedRelation) lfirst(lc);
+		queryEnv->namedRelList = list_delete(queryEnv->namedRelList, enr);
+		pfree(enr->md.name);
+		pfree(enr);
+	}
+	MemoryContextSwitchTo(oldcxt);
+}
+
+/* This is used when rolling back a subtransaction. */
+/* It is similar to ENRRollback */
+void ENRRollbackSubtransaction(SubTransactionId subid, QueryEnvironment *queryEnv)
+{
+	ListCell 	*lc;
+	List *enrs_to_drop = NIL;
+	MemoryContext oldcxt;
+
+	if (!queryEnv)
+		return;
+
+	oldcxt = MemoryContextSwitchTo(queryEnv->memctx);
+
+	/* Add back any ENRs that must be restored. */
+	foreach (lc, queryEnv->dropped_namedRelList)
+	{
+		EphemeralNamedRelation enr = (EphemeralNamedRelation) lfirst(lc);
+
+		Assert(enr->md.enrtype == ENR_TSQL_TEMP);
+		Assert(!enr->md.is_table_variable);
+
+		/* If this was dropped in the current subtransaction, then we must restore it. */
+		if (enr->md.dropped_subid != InvalidSubTransactionId && enr->md.dropped_subid >= subid) // 4
```

Comment:
how is dropped subid work here? we have 1, 2... 10 drops in 10 savepoints, how will we know up to which point we want to cleanup with single dropped subid?

Reply:
There is no way for an ENR to have its dropped subid altered. If the ENR is dropped in a subtransaction, then the subxactid is recorded, and the ENR is moved into the dropped_relid list - this is effectively the same as actually dropping it, but the information sticks around until commit/rollback. If the ENR is later recreated (ie, same name), a **new ENR is created**. Then if that new ENR is dropped, it is moved into the dropped_relid list with the new subtransaction id in which it was created/dropped.

Otherwise, if the same ENR is added back to namedRelList as part of rollback, you can see that the dropped_subid is reset. Then if it is dropped in a later subtransaction, it will be updated accordingly. 

Let me know if I misunderstood the question